drivers/led/lp50xx.c

```diff
@@ -0,0 +1,475 @@
+	LP5030,
+	LP5036,
+};
+
+struct lp50xx_config {
+	struct i2c_dt_spec bus;
+	const struct gpio_dt_spec gpio_enable;
+	enum lp50xx_chips chip;
+	uint8_t max_leds;
+	uint8_t num_leds;
+	uint8_t num_channels;
+	bool log_scale_en;
+	bool max_curr_opt;
+	uint8_t reset_reg;
+	uint8_t brightness_base_reg;
+	uint8_t color_base_reg;
+	const struct led_info *leds_info;
+};
+
+struct lp50xx_data {
+	struct led_data dev_data;
+	uint8_t *chan_buf;
+};
+
+static const struct led_info *lp50xx_led_to_info(
+			const struct lp50xx_config *config, uint32_t led)
+{
+	if (led < config->num_leds) {
+		return &config->leds_info[led];
+	}
+	return NULL;
+}
+
+static int lp50xx_get_info(const struct device *dev, uint32_t led,
+			const struct led_info **info)
+{
+	const struct lp50xx_config *config = dev->config;
+	const struct led_info *led_info = lp50xx_led_to_info(config, led);
+
+	if (!led_info) {
+		return -EINVAL;
+	}
+
+	*info = led_info;
+
+	return 0;
+}
+
+static int lp50xx_set_brightness(const struct device *dev,
+				 uint32_t led, uint8_t value)
+{
+	const struct lp50xx_config *config = dev->config;
+	const struct led_info *led_info = lp50xx_led_to_info(config, led);
+	struct lp50xx_data *data = dev->data;
+	struct led_data *dev_data = &data->dev_data;
+	uint8_t buf[2];
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (!IN_RANGE(value,
+		      dev_data->min_brightness,
+		      dev_data->max_brightness)) {
+		LOG_ERR("%s: brightness value out of bounds: "
+			"val=%d, min=%d, max=%d",
+			dev->name,
+			value,
+			dev_data->min_brightness,
+			dev_data->max_brightness);
+		return -EINVAL;
+	}
+
+	buf[0] = config->brightness_base_reg + led_info->index;
+	buf[1] = (value * 0xff) / 100;
+
+	return i2c_write_dt(&config->bus, buf, sizeof(buf));
+}
+
+static int lp50xx_on(const struct device *dev, uint32_t led)
+{
+	return lp50xx_set_brightness(dev, led, 100);
+}
+
+static int lp50xx_off(const struct device *dev, uint32_t led)
+{
+	return lp50xx_set_brightness(dev, led, 0);
+}
+
+static int lp50xx_set_color(const struct device *dev, uint32_t led,
+			    uint8_t num_colors, const uint8_t *color)
+{
+	const struct lp50xx_config *config = dev->config;
+	const struct led_info *led_info = lp50xx_led_to_info(config, led);
+	uint8_t buf[4];
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (num_colors != led_info->num_colors) {
+		LOG_ERR("%s: invalid number of colors: got=%d, expected=%d",
+			dev->name,
+			num_colors,
+			led_info->num_colors);
+		return -EINVAL;
+	}
+
+	buf[0] = config->color_base_reg + 3 * led_info->index;
+	buf[1] = color[0];
+	buf[2] = color[1];
+	buf[3] = color[2];
+
+	return i2c_write_dt(&config->bus, buf, sizeof(buf));
+}
+
+static int lp50xx_write_channels(const struct device *dev,
+				 uint32_t start_channel,
+				 uint32_t num_channels, const uint8_t *buf)
+{
+	const struct lp50xx_config *config = dev->config;
+	struct lp50xx_data *data = dev->data;
+	uint8_t end_channel, max_channels;
+
+	end_channel = start_channel + num_channels;
```

Comment:
OK, I saw the `end_channel` and fixed it but thanks for the tip.
For the first point, it should not be a problem because I compare the value with `max_channels` which is also zero-based so should be `- 1`

Reply:
Yes, works as intended and not a showstopper: it is only confusing when a variable contains a wrong value.